examples/all-clusters-app/esp32/main/AppTask.cpp

```diff
@@ -0,0 +1,178 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "AppTask.h"
+#include "DeviceWithDisplay.h"
+#include "Globals.h"
+#include "ScreenManager.h"
+#include "driver/gpio.h"
+#include "esp_log.h"
+#include "esp_spi_flash.h"
+#include "freertos/FreeRTOS.h"
+
+#define APP_TASK_NAME "APP"
+#define APP_EVENT_QUEUE_SIZE 10
+#define APP_TASK_STACK_SIZE (3072)
+
+static const char * TAG = "app-task";
+
+namespace {
+
+QueueHandle_t sAppEventQueue;
+TaskHandle_t sAppTaskHandle;
+
+} // namespace
+
+AppTask AppTask::sAppTask;
+
+CHIP_ERROR AppTask::StartAppTask()
+{
+    sAppEventQueue = xQueueCreate(APP_EVENT_QUEUE_SIZE, sizeof(AppEvent));
+    if (sAppEventQueue == NULL)
+    {
+        ESP_LOGE(TAG, "Failed to allocate app event queue");
+        return APP_ERROR_EVENT_QUEUE_FAILED;
+    }
+
+    // Start App task.
+    BaseType_t xReturned;
+    xReturned = xTaskCreate(AppTaskMain, APP_TASK_NAME, APP_TASK_STACK_SIZE, NULL, 1, &sAppTaskHandle);
+    return (xReturned == pdPASS) ? CHIP_NO_ERROR : APP_ERROR_CREATE_TASK_FAILED;
+}
+
+CHIP_ERROR AppTask::Init()
+{
+    /* Print chip information */
+    esp_chip_info_t chip_info;
+    esp_chip_info(&chip_info);
+    ESP_LOGI(TAG, "This is ESP32 chip with %d CPU cores, WiFi%s%s, ", chip_info.cores,
+             (chip_info.features & CHIP_FEATURE_BT) ? "/BT" : "", (chip_info.features & CHIP_FEATURE_BLE) ? "/BLE" : "");
+    ESP_LOGI(TAG, "silicon revision %d, ", chip_info.revision);
+    ESP_LOGI(TAG, "%dMB %s flash\n", spi_flash_get_chip_size() / (1024 * 1024),
+             (chip_info.features & CHIP_FEATURE_EMB_FLASH) ? "embedded" : "external");
+
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    statusLED1.Init(STATUS_LED_GPIO_NUM);
+    // Our second LED doesn't map to any physical LEDs so far, just to virtual
+    // "LED"s on devices with screens.
+    statusLED2.Init(GPIO_NUM_MAX);
+    bluetoothLED.Init();
+    wifiLED.Init();
+    pairingWindowLED.Init();
+
+    // Print QR Code URL
+    PrintOnboardingCodes(chip::RendezvousInformationFlags(CONFIG_RENDEZVOUS_MODE));
+
+#if CONFIG_HAVE_DISPLAY
+    InitDeviceDisplay();
+#endif
+    return err;
+}
+
+void AppTask::AppTaskMain(void * pvParameter)
+{
+    AppEvent event;
+    CHIP_ERROR err = sAppTask.Init();
+    if (err != CHIP_NO_ERROR)
+    {
+        ESP_LOGI(TAG, "AppTask.Init() failed due to %" CHIP_ERROR_FORMAT, err.Format());
+        return;
+    }
+
+    ESP_LOGI(TAG, "App Task started");
+
+    while (true)
+    {
+        BaseType_t eventReceived = xQueueReceive(sAppEventQueue, &event, pdMS_TO_TICKS(10));
+        while (eventReceived == pdTRUE)
+        {
+            sAppTask.DispatchEvent(&event);
+            eventReceived = xQueueReceive(sAppEventQueue, &event, 0);
```

Comment:
What does 0 stand for? I was hoping it would be something like waiting indefinitely:

```
while (xQueueReceive(sAppEventQueue, &event, portMAX_DELAY) {
Handle/Process button press
}
```

Reply:
I add a comment on it.0 means return immediately if the queue is empty